### PR TITLE
Small code and logging improvements

### PR DESCRIPTION
**Pull request by rogierslag**
@inventid/tech 

I've added some logging and stuff, and moved a bit around to make deployments easier. Can you have a look?

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -16,45 +16,92 @@ var nightmare = Nightmare(),
...
-// parse json responses.
-app.use(bodyParser.json());
-
-app.post('/render', function(req, res) {
-  queue.add(function() {
-    return render(req.body)
-  }).then(function(output) {
-    res.sendFile(output);
-  });
-});
```
Comment (rogierslag) [inline index.js:29]: this was added to `startServer` below

File: index.js [hunk 2]
```diff
@@ -16,45 +16,92 @@ var nightmare = Nightmare(),
...
+    router.use(express.static(config.document_path));
+    router.get('/data.json', function (req, res) {
+        log('debug', 'Serving data.json for id ' + id);
+        res.json(data);
+    });
+    router.get('/data', function (req, res) {
+        log('debug', 'Serving data in javascript for id ' + id);
+        res.type("text/javascript").send("var data = " + JSON.stringify(data));
+    });
+    router.get('/defer', function (req, res) {
```
Comment (rogierslag) [inline index.js:38]: Can you elaborate in comments when this is fired?

File: index.js [hunk 3]
```diff
@@ -16,45 +16,92 @@ var nightmare = Nightmare(),
...
+        deferred = new Promise(function (resolve, reject) {
```
Comment (rogierslag) [inline index.js:39]: And why its safe to depend on this var in the promise below?

File: index.js [hunk 4]
```diff
@@ -16,45 +16,92 @@ var nightmare = Nightmare(),
...
+            router.get('/done', function (req, res) {
+                log('debug', 'Document called done. id: ' + id);
+                resolve();
```
Comment (rogierslag) [inline index.js:42]: and what this callback contains?

File: index.js [hunk 5]
```diff
@@ -16,45 +16,92 @@ var nightmare = Nightmare(),
...
         res.status(200).end();
-      });
     });
 
-    res.status(200).end();
-  });
+    return Promise
+        .resolve(nightmare.goto('http://localhost:' + getMyPort() + '/' + id))
+        .then(function () {
+            return deferred;
```
Comment (rogierslag) [inline index.js:53]: I dont know what happens here. This can sometimes be `null` right? Or am I missing something?

File: index.js [hunk 6]
```diff
@@ -16,45 +16,92 @@ var nightmare = Nightmare(),
...
+        })
+        .then(function () {
+            // Converting the HTML to PDF with nightmare
+            log('debug', 'Outputting the html to pdf');
+            return nightmare.pdf(output, {marginsType: 2});
+        })
+        .then(function () {
+            // Return the location of the document
+            return output;
```
Comment (rogierslag) [inline index.js:62]: shouldnt we delete this later on?

File: index.js [hunk 7]
```diff
@@ -16,45 +16,92 @@ var nightmare = Nightmare(),
...
+    });
+    app.post('/render', function (req, res) {
+        log('info', 'Received rendering request');
+        queue.add(function started() {
+            log('info', 'Started rendering request');
+            return render(req.body)
+        }).then(function finished(output) {
+            log('info', 'Finished rendering request');
+            res.sendFile(output);
+        });
```
Comment (rogierslag) [inline index.js:104]: https://github.com/inventid/gennie/pull/5/files?diff=unified#r42151624 like here?

File: index.js [hunk 8]
```diff
@@ -16,45 +16,92 @@ var nightmare = Nightmare(),
...
+
+// Log some stuff to the console
+function log(level, message) {
+    console.log("[" + level.toUpperCase() + "]: " + message);
+}
+
+// Ge the servers configured port (either config file or environment variable PORT)
+function getMyPort(withLog) {
+    var port = config.port;
+    if (process.env.PORT) {
```
Comment (rogierslag) [inline index.js:74]: by using PORT to override the port number, this makes it very easy for docker deployments :)

**Comment by rogierslag** [comment]
Performance test: 1000 pdf were generated no problem whatsoever in 15min35sec within Docker :)

File: index.js (hunk 3 above)
Comment (joostverdoorn) [inline index.js:39]: `deferred` is either undefined or a promise. When it's undefined, `.then(deferred)` instantly resolves. When it's a promise,  `.then(deferred)` resolves when `deferred` resolves.

File: index.js (hunk 5 above)
Comment (joostverdoorn) [inline index.js:53]: `deferred` is either undefined or a promise. When it's undefined, `.then(deferred)` instantly resolves. When it's a promise,  `.then(deferred)` resolves when `deferred` resolves.

File: index.js (hunk 6 above)
Comment (joostverdoorn) [inline index.js:62]: Should we?

File: index.js (hunk 7 above)
Comment (joostverdoorn) [inline index.js:104]: Sure

File: index.js [hunk 9]
```diff
@@ -16,45 +16,92 @@ var nightmare = Nightmare(),
...
+        if (withLog) {
+            log('warning', 'Environment variable PORT will take precedence over the port in the config');
+        }
+        port = process.env.PORT;
+    }
+    return port;
+}
+
+// Start the server
+function startServer() {
```
Comment (joostverdoorn) [inline index.js]: Why put this in a separate function?

File: index.js [hunk 10]
```diff
@@ -16,45 +16,92 @@ var nightmare = Nightmare(),
...
+        });
+}
+
+// Log some stuff to the console
+function log(level, message) {
+    console.log("[" + level.toUpperCase() + "]: " + message);
+}
+
+// Ge the servers configured port (either config file or environment variable PORT)
+function getMyPort(withLog) {
```
Comment (joostverdoorn) [inline index.js:72]: Why put it in a separate function?

File: index.js [hunk 11]
```diff
@@ -1,11 +1,11 @@
...
+    Promise = require('bluebird'),
+    Queue = require('promise-queue'),
+    express = require('express'),
     bodyParser = require('body-parser'),
-    uuid       = require('node-uuid'),
-    config     = require('./config.json'),
-    app        = express();
+    uuid = require('node-uuid'),
+    config = require('./config.json'),
+    app = express();
```
Comment (joostverdoorn) [inline index.js:10]: I don't think this is better

File: index.js (hunk 10 above)
Comment (rogierslag) [inline index.js:72]: so I can reuse it (for the server startup and the location where we send nightmare to the server)

File: index.js (hunk 11 above)
Comment (rogierslag) [inline index.js:10]: my IDE autoformatted this, will put it back!

File: index.js (hunk 10 above)
Comment (joostverdoorn) [inline index.js:72]: I would say we calculate and store it once, so we don't have to do that for each render!

File: index.js (hunk 3 above)
Comment (rogierslag) [inline index.js:39]: :+1:

File: index.js (hunk 6 above)
Comment (rogierslag) [inline index.js:62]: I meant the file itself after serving it

**Comment by rogierslag** [comment]
I fixed the first round of feedback :)

File: index.js (hunk 6 above)
Comment (rogierslag) [inline index.js:62]: Which is now done a line 102

**Comment by joostverdoorn** [comment]
Looks good! :shipit: :+1:

**Comment by joostverdoorn** [comment]
btw, how am I the author of that commit?

**Comment by rogierslag** [comment]
Because I just moved some code around and felt that that would inadvertently take over your entire code. That didnt seem reasonable

**Comment by joostverdoorn** [comment]
There, I fixed it for you. Although I appreciate the sentiment, I would prefer if you didn't mess around with authorship ;)

Now LGTM

**Comment by rogierslag** [comment]
:+1:

**Pull request merged**